### PR TITLE
Throw an exception on unknown trait/weapon fields.

### LINKED ISSUE #15201: [Suggestion] - Force the mod to crash at startup when inputting invalid propierties on traits.
Im referencing my comment in #9810 , as an user who has to update mods constantly, I believe this will solve the issues with renaming properties of traits.

### PR DESCRIPTION
**Pull request by pchote**
Fixes #15201. This also reports the errors in `--check-yaml`.

~~Depends on #15252 and #15263.  Only the last two commits are new.~~

### CODE REVIEW

**Comment by reaperrr** [comment]
Both dependencies have been merged.

**Comment by pchote** [comment]
Rebased.

File: OpenRA.Game/GameRules/WeaponInfo.cs [hunk 1]
```diff
@@ -86,12 +86,22 @@ public sealed class WeaponInfo
...
 			content.Nodes = MiniYaml.Merge(new[] { content.Nodes });
-			FieldLoader.Load(this, content);
+
+			try
+			{
+				FieldLoader.Load(this, content, ThrowOnUnknownExceptWarheadsOrProjectiles);
+			}
+			catch (FieldLoader.UnknownFieldsException e)
+			{
+				throw new YamlException("Weapon name " + name + "; " + e.Message);
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Game/GameRules/WeaponInfo.cs]: This ends up like "Weapon name <myweapon>; Projectile name <myprojectile>: blah".  I  felt that `;` read better than `:` in this situation.

**Comment by GraionDilach** [comment]
Could we not crash the Utility? Crashing the base game is alright to me, but crashing the Utility every time on each take is tedious.

**Comment by pchote** [comment]
Can you please provide a testcase? `--check-yaml` should be listing errors, not crashing.

**Comment by GraionDilach** [comment]
Adding this to the RA defaults.yaml

```
^ExistsInWorld:
	AppearsOnRadar:
	CombatDebugOverlay:
	GivesExperience:
		PlayerExperienceModifier: 1
		CanMakeStuffUp: true
	ScriptTriggers:
	ConditionManager:
	RenderDebugState:
```

gave me

```
d:\OpenRA-PRcheck>OpenRA.Utility ra --check-yaml
Testing mod: Red Alert
OpenRA.Utility(1,1): Error: Failed with exception: System.AggregateException: One or more errors occurred. ---> OpenRA.YamlException: Actor type fact.colorpicker: Trait name GivesExperience: Unknown property defined: CanMakeStuffUp
   at OpenRA.ActorInfo..ctor(ObjectCreator creator, String name, MiniYaml node) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\ActorInfo.cs:line 63
   at OpenRA.Ruleset.<>c__DisplayClasse.<LoadDefaults>b__7(MiniYamlNode k) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 132
   at OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, String debugName, Func`2 logKey, Func`2 logValue) in d:\OpenRA-PRcheck\OpenRA.Game\Exts.cs:line 376
   at OpenRA.Ruleset.MergeOrDefault[T](String name, IReadOnlyFileSystem fileSystem, IEnumerable`1 files, MiniYaml additional, IReadOnlyDictionary`2 defaults, Func`2 makeObject, Func`2 filterNode) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 120
   at OpenRA.Ruleset.<>c__DisplayClasse.<LoadDefaults>b__6() in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 131
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout)
   at OpenRA.Ruleset.LoadDefaults(ModData modData) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 162
   at OpenRA.ModData.<.ctor>b__3() in d:\OpenRA-PRcheck\OpenRA.Game\ModData.cs:line 95
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at OpenRA.ModData.get_DefaultRules() in d:\OpenRA-PRcheck\OpenRA.Game\ModData.cs:line 41
   at OpenRA.Mods.Common.UtilityCommands.CheckYaml.OpenRA.IUtilityCommand.Run(Utility utility, String[] args) in d:\OpenRA-PRcheck\OpenRA.Mods.Common\UtilityCommands\CheckYaml.cs:line 64
---> (Inner Exception #0) OpenRA.YamlException: Actor type fact.colorpicker: Trait name GivesExperience: Unknown property defined: CanMakeStuffUp
   at OpenRA.ActorInfo..ctor(ObjectCreator creator, String name, MiniYaml node) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\ActorInfo.cs:line 63
   at OpenRA.Ruleset.<>c__DisplayClasse.<LoadDefaults>b__7(MiniYamlNode k) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 132
   at OpenRA.Exts.ToDictionaryWithConflictLog[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, String debugName, Func`2 logKey, Func`2 logValue) in d:\OpenRA-PRcheck\OpenRA.Game\Exts.cs:line 376
   at OpenRA.Ruleset.MergeOrDefault[T](String name, IReadOnlyFileSystem fileSystem, IEnumerable`1 files, MiniYaml additional, IReadOnlyDictionary`2 defaults, Func`2 makeObject, Func`2 filterNode) in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 120
   at OpenRA.Ruleset.<>c__DisplayClasse.<LoadDefaults>b__6() in d:\OpenRA-PRcheck\OpenRA.Game\GameRules\Ruleset.cs:line 131
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---
```

**Comment by pchote** [comment]
> OpenRA.Utility(1,1): Error: Failed with exception:

This is the utility reporting the exception as expected.  The stacktrace is provided to help debugging, the same as e.g. missing dependencies.

**Comment by GraionDilach** [comment]
Yeah, but you need to restart the game or the Utility after every fixed lingering property - I am really expecting there will be a lot of these in some projects which would greatly benefit if the utility would catch these YamlExceptions and output their text.

**Comment by pchote** [comment]
Yes, this is the same as for missing dependencies.  The engine does not support that for any utility / lint checks, and fixing that is way out of scope of this PR.

**Comment by GraionDilach** [comment]
Well, I definitely will not approve this in this form for sure. This is a PITA to work with as it is.

**Comment by pchote** [comment]
Ok.  Hopefully someone can resurrect this in the future then.

**Comment by GraionDilach** [comment]
I am currently playing with defining a static Game.IsUtility boolean which I would incorporate into ThrowOnAllUnknownFields as a solution to the problem.

**Comment by GraionDilach** [comment]
@pchote Would you consider something along the path of https://github.com/GraionDilach/OpenRA/commit/1eafad6eb8bdbac6a57f005d3b5aaa3baa5e5941 as a solution?

**Pull request closed without merging**